### ISSUE TITLE
[Enhancement] Bordered views, e.g. grid view

### ISSUE THREAD

**Issue by PureWeen** [xamarin/Xamarin.Forms#1695]
## Rationale

Views can't easily be given borders without nesting them in frames or implementing custom renderers. 

## Implementation

```csharp
public static readonly BindableProperty BorderThicknessProperty =
     BindableProperty.Create(propertyName: nameof(BorderThickness),
	     returnType: typeof(float), declaringType: typeof(Image));
		
public static readonly BindableProperty BorderColorProperty =
    BindableProperty.Create(propertyName: nameof(BorderColor),
		returnType: typeof(Color), declaringType: typeof(Image));

public static readonly BindableProperty BorderFillColorProperty =
	BindableProperty.Create(propertyName: nameof(BorderFillColor),
		returnType: typeof(Color), declaringType: typeof(Image));
```

Renderers will frame the given view using the given border properties

### Android
- `View.BorderThickness`:  should set the thickness of the border
- `View.BorderColor`:  should set the outline colors of the border
- `View.BorderFillColor`:  should set the internal fill on the border. If fill color is different than Border Color you'll have   | Border Color | Fill Color | Border Color | Component

### iOS
See Android

### UWP
See Android

## Implications for CSS
```css
{border-color: coral;}
{border-width: thin;}

// Fill color possible with addition of pseudo elements
entry::borderfill
{
color
}
```
## Backward Compatibility
This is adding a new set of properties so should be backwards compatible. Need to ensure that if the user doesn't specify any of these properties that the entry field appears normal

## Difficulty : Easy

This issue may correlate with implementing following issues: 
https://github.com/xamarin/Xamarin.Forms/issues/1664
https://github.com/xamarin/Xamarin.Forms/issues/1681

**Comment by andreinitescu** [comment]
This would become extremely handy. Is `View` the right class to have this on? I guess so... 
I was thinking about ListView cells which derive from `Element` and they won't get this.

**Comment by andreinitescu** [comment]
@jassmith @davidortinau It's not clear to me how this should correlate with #1664 and #1681. There are people who started working on those already. How is this going to work with those?

**Comment by PureWeen** [comment]
As far as the correlation at a minimum I just wanted to make sure the apis between the three stayed consistent.  There's an IBorderElement interface that was used for CSS but I was thinking we'd add the two additional properties to that so there's a consistent spec among the elements that can specify Border properties.  

Then at the platform level I just wanted the three tickets to be aware of each other because I feel like there will be consistent code between them with how borders are implemented

**Comment by hartez** [comment]
> If fill color is different than Border Color you'll have | Border Color | Fill Color | Border Color | Component

@PureWeen how do we determine the thickness of the Border Color portion of that diagram?

**Comment by hartez** [comment]
> There's an IBorderElement interface that was used for CSS but I was thinking we'd add the two additional properties to that so there's a consistent spec among the elements that can specify Border properties.

@PureWeen Can you update the spec to make that explicit (that these properties are being added to IBorderElement and the BorderElement static class)?

**Comment by andreinitescu** [comment]
I was wondering why do interfaces like `IBorderElement` or `IPaddingElement` exist instead of just having properties?  
I don't see any implementation currently in XF which needs to get hold of a `IBorderElement` instance for some specific purpose.

**Comment by hartez** [comment]
Those interfaces enforce a set of properties which an Element requires for a particular set of features. This can be used to map CSS properties (take a look at AssemblyInfo.cs in the Core project); the properties can be mapped to `IPaddingElement` which applies to Frame, Layout, and Page right now; anything else we later mark with that interface will also support those properties.

It also gives us an easy way to write some generic methods for handling groups of properties in other places; take a look at how IFontElement is used in the platform projects.

Finally, it goes a little way to ensuring that we have consistent property names and types across the Elements.

**Comment by PureWeen** [comment]
@hartez   good point added an extra property for that and some additional consideration notes

**Comment by hartez** [comment]
@PureWeen What's the rationale behind having an internal color in a Border? What are we gaining with that extra layer? When I think about adding a border to an element, I usually think of just a solid (single color) outline of a specified thickness.

**Comment by PureWeen** [comment]
@hartez   that mainly came from looking around at how/what kinds of borders I saw.  A very typical use case I saw would be to have a black outline on the border. Especially in cases with rounded images where it would be
Black | Color | Black 

So mainly just trying to capture a larger set of the use case as I can with the API before someone has to use a renderer

But I could see the fill being split out to a feature and then just seeing if it gets traction

**Comment by andreinitescu** [comment]
@PureWeen  I've personally never used border fill. How is it implemented natively exactly? Can you please give a very quick example? Any platform will do

**Comment by PureWeen** [comment]
```
<Border>
<Border>
<Border>
```

But it probably doesn't make since as a parameter and more interesting borders should be implemented through a different mechanism. For example just nesting things that have borders since after this everything can have borders

**Comment by StephaneDelcroix** [comment]
see also #1754

**Comment by andreinitescu** [comment]
@PureWeen Personally I'm not in favor of the border filling idea. But I see you already edited the ticket and remove that 👍

**Comment by PureWeen** [comment]
As @StephaneDelcroix  pointed out there's some related internal discussion/work going on in https://github.com/xamarin/Xamarin.Forms/issues/175. We're working out the best way to expose all of this in a simple/useful way. Once those ideas come together we'll update these sets of tickets with more thoughts on implementation details

**Comment by samhouts** [comment]
Related to #1814

**Comment by StephaneDelcroix** [comment]
Related to #1754

**Comment by PureWeen** [comment]
Some work I'm doing currently related to borders
https://github.com/xamarin/Xamarin.Forms/issues/1724

I realize this ticket expands on the concept of borders but I just wanted to make whoever works on this aware

On Android I've generalized
https://github.com/xamarin/Xamarin.Forms/blob/master/Xamarin.Forms.Platform.Android/ButtonBackgroundTracker.cs

so it can essentially be applied to any Android.Views.View but it uses the Button API currently which we probably wanted to expand.   Just wanted to make a note of other current efforts in Border-land

**Comment by Mrnikbobjeff** [comment]
Is this currently ready to implement as in the proposal? I would like to work in this if this is the case.

**Comment by samhouts** [comment]
@Mrnikbobjeff Yes, it is ready! We'll reserve this issue for you. Thank you!

**Comment by samhouts** [comment]
@Mrnikbobjeff Were you able to create a PR for this issue? Thanks!

**Comment by Happypig375** [comment]
Please add Editor to the list. Currently, Editor has a border, at least on UWP.

![image](https://user-images.githubusercontent.com/19922066/59548993-57552480-8f89-11e9-8357-23e4b9c72916.png)

**Comment by adrianknight89** [comment]
Can we add the ability to define which sides of a view will have a border? I have at least one case where I need to add a top and bottom border and ignore the horizontal sides. The view spans across the device screen horizontally, so it doesn't make sense to add a left/right border.

**Comment by Happypig375** [comment]
You can set left, right, top and bottom separately for `Thickness`.

**Comment by adrianknight89** [comment]
@Happypig375 My bad. You're right! I must have been thinking of something else last night. Hopefully, `Thickness` will be tested properly so that a `0` on any side will not create dead space.

**Comment by titoleru** [comment]
Hi @samhouts, is it planned to extend this to Editors? There is an unexpected behavior of iOS Editor, which does not show any borders, unlike Android or the entries do by default. Right now this can only be fixed by using a custom renderer
